File: LC/Python/Medium/3573.py
class Solution:
    def maximumProfit(self, prices: List[int], k: int) -> int:
        n = len(prices)
        cache = {}

        # state: 0 = flat, 1 = long, 2 = short
        # transactions = completed transactions so far
        def dfs(i, transactions, state):
            if (i, transactions, state) in cache:
                return cache[(i, transactions, state)]
            if transactions == k:
                return 0 if state == 0 else float("-inf")
            if i == n:
                return 0 if state == 0 else float("-inf")

            p = prices[i]
            if state == 0:
                res = max(
                    dfs(i + 1, transactions, 0),
                    dfs(i + 1, transactions, 1) - p,
                    dfs(i + 1, transactions, 2) + p
                )
            elif state == 1:
                res = max(
                    dfs(i + 1, transactions, 1),
                    dfs(i + 1, transactions + 1, 0) + p
                )
            else:
                res = max(
                    dfs(i + 1, transactions, 2),
                    dfs(i + 1, transactions + 1, 0) - p
                )

            cache[(i, transactions, state)] = res
            return res

        return dfs(0, 0, 0)

File: LC/C++/Medium/3573.cpp
class Solution {
    vector<vector<vector<long long>>> cache_;
public:
    long long dfs(vector<int>& prices, int i, int k, int state) {
        if (i == prices.size()) {
            return state == 0 ? 0 : LLONG_MIN / 4;
        }
        if (cache_[i][k][state] != LLONG_MIN) {
            return cache_[i][k][state];
        }

        int p = prices[i];
        long long res = dfs(prices, i + 1, k, state);
        if (k > 0 && state == 0) {
            res = max(res, dfs(prices, i + 1, k - 1, 1) - p);
            res = max(res, dfs(prices, i + 1, k - 1, 2) + p);
        } else if (state == 1) {
            res = max(res, dfs(prices, i + 1, k, 0) + p);
        } else {
            res = max(res, dfs(prices, i + 1, k, 0) - p);
        }

        return cache_[i][k][state] = res;
    }

    long long maximumProfit(vector<int>& prices, int k) {
        cache_.assign(
            prices.size(),
            vector<vector<long long>>(
                k + 1,
                {LLONG_MIN, LLONG_MIN, LLONG_MIN}
            )
        );
        return dfs(prices, 0, k, 0);
    }
};